### PR TITLE
Fix the OpenshiftClient connection leaks

### PR DESCRIPTION
**Pull request by davidfestal**
### What does this PR do?

Fix the OpenshiftClient connection leaks ... while reusing a common `OkHttpClient` instance as much as possible (not possible in some cases, like `startExec()`, that involve
`InputStreamPumper` and broken pipes).

### What issues does this PR fix or reference?

This PR is related to the issue discussed [here](https://github.com/eclipse/che/issues/5902#issuecomment-346858757) and [here](https://github.com/eclipse/che/issues/5902#issuecomment-348516145), as well as this initial issue of the `OpenshiftConnector` : https://issues.jboss.org/browse/CHE-180

It fixes these issues in a way that is compatible with the CHE-5-based OSIO multi-tenant server.

### CODE REVIEW

File: plugins/plugin-docker/che-plugin-openshift-client/src/main/java/org/eclipse/che/plugin/openshift/client/OpenShiftClientFactory.java [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+@Singleton
+public class OpenShiftClientFactory {
+
+  private OkHttpClient httpClient;
+
+  @Inject
+  public OpenShiftClientFactory(
+      OpenshiftWorkspaceEnvironmentProvider workspaceEnvironmentProvider) {
+    this.httpClient =
+        HttpClientUtils.createHttpClient(workspaceEnvironmentProvider.getDefaultOpenshiftConfig());
```
Comment (ibuziuk) [inline plugins/plugin-docker/che-plugin-openshift-client/src/main/java/org/eclipse/che/plugin/openshift/client/OpenShiftClientFactory.java:39]: @davidfestal not sure, but formatting does not look right from the first glance

Comment (davidfestal) [reply plugins/plugin-docker/che-plugin-openshift-client/src/main/java/org/eclipse/che/plugin/openshift/client/OpenShiftClientFactory.java:39]: well, it's formatted by the maven formatter plugin afaik

Comment (ibuziuk) [reply plugins/plugin-docker/che-plugin-openshift-client/src/main/java/org/eclipse/che/plugin/openshift/client/OpenShiftClientFactory.java:39]: ok, cool - sorry than ;-)

**Comment by benoitf** [comment]
ci-build

File: plugins/plugin-docker/che-plugin-openshift-client/src/main/java/org/eclipse/che/plugin/openshift/client/OpenShiftClientFactory.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+  }
+
+  @PreDestroy
+  public void cleanup() {
+    if (httpClient.connectionPool() != null) {
```
Comment (sleshchenko) [inline plugins/plugin-docker/che-plugin-openshift-client/src/main/java/org/eclipse/che/plugin/openshift/client/OpenShiftClientFactory.java]: As far as I see checking `httpClient.connectionPool()` on null is redundant at least for 3.6.0 http client https://github.com/square/okhttp/blob/parent-3.6.0/okhttp/src/main/java/okhttp3/OkHttpClient.java#L738. The same for `httpClient.dispatcher()` and `httpClient.dispatcher().executorService()`
Of course up to you to leave these checks or not.

**Comment by davidfestal** [comment]
ci-build

**Comment by benoitf** [comment]
ci-build

**Comment by benoitf** [comment]
ci-build

**Review by l0rd — APPROVED** [review]

**Review by ibuziuk — APPROVED** [review]

**Pull request merged**